sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1950,522 @@ def transform(self, X):
+            needs to be nonnegative.
+        """
+        n_samples, n_features = X.shape
+
+        # for compatibility issue with numpy<=1.8.X, references
+        # need to be a list scaled between 0 and 100
+        references = list(map(lambda x: x * 100, self.references_))
+        self.quantiles_ = []
+        for feature_idx in range(n_features):
+            column_nnz_data = X.data[X.indptr[feature_idx]:
+                                     X.indptr[feature_idx + 1]]
+            if len(column_nnz_data) > self.subsample:
+                column_subsample = (self.subsample * len(column_nnz_data) //
+                                    n_samples)
+                if self.ignore_implicit_zeros:
+                    column_data = np.zeros(shape=column_subsample,
+                                           dtype=X.dtype)
+                else:
+                    column_data = np.zeros(shape=self.subsample, dtype=X.dtype)
+                column_data[:column_subsample] = choice(
+                    column_nnz_data, size=column_subsample,
+                    replace=False, random_state=random_state)
+            else:
+                if self.ignore_implicit_zeros:
+                    column_data = np.zeros(shape=len(column_nnz_data),
+                                           dtype=X.dtype)
+                else:
+                    column_data = np.zeros(shape=n_samples, dtype=X.dtype)
+                column_data[:len(column_nnz_data)] = column_nnz_data
+
+            if not column_data.size:
+                # if no nnz, an error will be raised for computing the
+                # quantiles. Force the quantiles to be zeros.
+                self.quantiles_.append([0] * len(references))
+            else:
+                self.quantiles_.append(
+                    self._compute_quantile_one_column(column_data, references,
+                                                      random_state))
+        self.quantiles_ = np.transpose(self.quantiles_)
+
+    def fit(self, X, y=None):
+        """Compute the quantiles used for transforming.
+
+        Parameters
+        ----------
+        X : ndarray or sparse matrix, shape (n_samples, n_features)
+            The data used to scale along the features axis. If a sparse
+            matrix is provided, it will be converted into a sparse
+            ``csc_matrix``. Additionally, the sparse matrix needs to be
+            nonnegative if `ignore_implicit_zeros` is False.
+
+        Returns
+        -------
+        self : object
+            Returns self
+        """
+        X = check_array(X, accept_sparse='csc')
+        rng = check_random_state(self.random_state)
+
+        if self.n_quantiles <= 0:
+            raise ValueError("Invalid value for 'n_quantiles': %d. "
+                             "The number of quantiles must be at least one."
+                             % self.n_quantiles)
+
+        if self.subsample <= 0:
```

Comment:
I was wondering if we should raise an error or just set `n_quantiles = subsample` in this case and raise a warning.

@jnothman WDYT?